sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -652,10 +652,11 @@ object DataSourceStrategy {
  */
 object PushableColumn {
   def unapply(e: Expression): Option[String] = {
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
     def helper(e: Expression) = e match {
       case a: Attribute => Some(a.name)
       case _ => None
     }
-    helper(e)
+    helper(e).map(quoteIfNeeded)
```

Comment:
OK this is a behavior change.

`Filter#references` is an `Array[String]` so there is no confusion if we won't quote. What's our plan to support nested fields?

Reply:
Each element in `Filter#references` is a column that this Filter refers to. There can be multiple of them. As a result, quoting on column name that contains `dot` is still needed.

Reply:
Ah sorry I misread it with the v2 reference.

So our plan is to use dot to separate nested fields in v1 reference. Is this a 3.1 feature or we plan to make it into 3.0?

Reply:
I would like to make it into 3.0 as this lays a good foundation to support features such as pushdown on column containing dots and nested predicate pushdown. Also, I feel it’s a good time to have a small breaking change in 3.0 instead of 3.1. WDYT?

Reply:
Is this really needed to support column name containing dot?

And a few thoughts to roll it out more smoothly:
1. if we refer to a top-level column, don't quote it (so no breaking change)
2. if we refer to a nested field, use a special encoding (like json array?)

If users use a v1 source with Spark 3.0, and Spark pushes down a filter with nested fields, then it would either fail or ignore it as it doesn't recognize the json array encoded column name.

However, if we use dot to separate field names, then v1 source can be broken if a filter is pushed down pointing to a top-level column containing dot.

Reply:
Implementation-wise, I think it's easier to use dot to separate field names, both for Spark and source developers. I don't know how common it is to see top-level column names containing dot.

cc @brkyvz @rdblue @imback82

Reply:
We support column names containing `dots` in ORC.

I feel special encoding makes the implementation really over complicated. I believe that currently, only ORC support column name containing `dots`, so the impact is not much.

Reply:
@dbtsai, I thought Parquet also supports the dots in column names at https://github.com/apache/spark/pull/27780.
Also, the problem is that there are multiple external sources. Dots can be used to express namespace-like annotations just like our SQL configurations.

JSON array might not be an option either because column names themselves can be JSON array:

```scala
scala> sql("""select 'a' as `["a", "'b"]`""")
res1: org.apache.spark.sql.DataFrame = [["a", "'b"]: string]
```

It's unlikely but it's possible breaking change.

The only option I can think of now is to have a SQL conf that lists datasources to support to backquote dots in column name (and don't backquote for nested column access).

Reply:
@HyukjinKwon You are right. I'm just breaking https://github.com/apache/spark/pull/27780 into smaller PR so it's easier to review. Chat with @cloud-fan offline, and I'll close this one and work on the other bigger one to avoid the confusion. Let's discuss how we want to handle the API compatibility in the other PR.

Reply:
Thank you @dbtsai!